### ISSUE TITLE
[FEATURE REQUEST]: TPE-model

### ISSUE THREAD

**Issue by NormanTUD** [pytorch/botorch#2807]
### Motivation

Before discovering ax/botorch, I used hyperopt. I like ax more, because it is way more modular. But still, for some problems, I miss TPE. How could TPE, like from hyperopt, be implemented in BoTorch? Actually, I must admit that I am just not smart enough to implement that, but I'd love it if it was available.

### Describe the solution you'd like to see implemented in BoTorch.

Similiar to using Botorch_modular or Uniform or Sobol, allow me to use TPE as a model. Would be greatly appreciated.

### Pull Request

None

### Code of Conduct

- [x] I agree to follow BoTorch's Code of Conduct

**Comment by NormanTUD** [comment]
Thanks! This seems quite promising, but I ran into many problems there. Is there any more complete example, for example, so that I can use `ax_client.to_json_snapshot()`? That one, for example, gave me an error that it had not found "registered with a corresponding encoder in ENCODER_REGISTRY", and I actually don't find any useful info on how to write custom Encoders. Same for other parameters.

I think having a working example that implements all of that would be of immense help, since it could be adapted to a lot of different use cases and generators.

**Comment by eytan** [comment]
Out of curiosity, what problems do you have where you’d prefer TPE over BO?

Note that (as Louis recommended), Ax is the preferred tool to use if you
are wanting to apply BO to a problem (whereas BoTorch is the framework
you’d use to implement new BO algorithms at a low level). Ax handles a lot
of tricky stuff with normalizing the data, encoding discrete variables
(categorical, integer, ordered categorical), and automatically selecting
the right acquisition function optimizers to use for your setup.

We recently ran some benchmarks and found Ax to consistently outperform
TPEs on all classes of problems so I’d be curious what problems you are
finding TPEs to be beneficial.

Best
E

**Comment by NormanTUD** [comment]
I work at a research center for AI, and we offer a tool for easy hyperparameter optimization. An earlier version of the tool used HyperOpt/TPE, and the newer one uses Botorch for finding good hyperparameters. Since the old version is now deprecated and we don't want to spend any more time developing it, and the newer one has lots of additional functionality, like MOO for example, we don't want to give people the old one anymore unless absolutely neccessary. 

One of our users got to me with the request to support TPE for Ax/BT as well. He used it earlier, and now wants to use our modern tools to compare what TPE can do vs. what Botorch can do. So, I don't really have a "real" use for it because its better or something. 

But I'd love to support it anyway, given that if we could properly implement it, we could expand that to have any external program handle generation of new points-to-be-evaluated. Like, the idea would be that the user can simply point to an external python-script (which, for example, could use TPE), which gets a JSON file as parameter (for example, having the parameter ranges, types and previous evaluations, and let that program output certain hyperparameters). This would make our tool massively more flexible. 

For this it would be amazing to have a fully working example, like with MOO, and `ENCODER_REGISTRY` (so that `to_json_snapshot` works) and so on.

Our tool, if you are interested, is free software. 

https://imageseg.scads.de/omniax/

Currently, it's hosted there. Example runs can be created here -> https://imageseg.scads.de/omniax/gui 

They are customized for our cluster, but should work on all Linux machines, and all kinds of clusters. It's still work in progress, though.

**Comment by eytan** [comment]
Adding TPEs as a generation strategy or external generation strategy in Ax
is the way to go. Also if you are providing a tool for HPO I’d recommend
calling Ax, it’s what we use internally at meta for 1000+ engineers and
researchers, and will work much better than using botorch directly for
general purpose HPO tasks.

It also doesn’t make sense to use BoTorch for TPEs-they don’t use
probabilistic models or gradient-based optimization. Perhaps you can open a
new issue / feature request on GH for TPE support in Ax and we can discuss
there?

**Comment by NormanTUD** [comment]
I am using ax, and not BT directly. But I thought since it's a model related issue, it would be better off to be posted here. But thanks, I will create a Feature Request in Ax's github.

**Comment by ltiao** [comment]
Thanks for opening https://github.com/facebook/Ax/issues/3612. Will close this one and continue the discussion there!
